src/rt/aaA.d

```diff
@@ -474,6 +474,16 @@ extern (C) void _aaClear(AA aa) pure nothrow
     }
 }
 
+/// Reserve AA
+extern (C) void _aaReserve(AA* aa, const TypeInfo_AssociativeArray ti, size_t ndim)
+{
+    // lazily alloc implementation
```

Comment:
The dimension has to be a power of 2, so anything else should be rejected or rounded up.

Reply:
Good catch, this actually saved me some headache I had with a few tests. Implemented in reserve(). Thanks!

Reply:
If the resize call is supposed to avoid rehashing for the given number of entries, you need to include the "grow ratio", too. The AA rehashes if the number of entries exceeds 4/5 of the dimension of the bucket array: https://github.com/Darredevil/druntime/blob/eb9865e63b1205e36f28c52aeffbb08d02f96b68/src/rt/aaA.d#L386.

Reply:
@rainers that seems to explain the efficiency issue. Also, @Darredevil what happens if reserve is called with a smaller number than the current number of elements? Seems to me the correct code should be:

```d
if (ndim <= aa.dim) return;
if (aa.used * GROW_DEN > ndim * GROW_NUM) ndim = (aa.used * GROW_DEN + GROW_NUM - 1) / GROW_NUM;
assert(aa.used * GROW_DEN <= ndim * GROW_NUM);
ndim = nextpow2(ndim);
if (aa.impl is null)
        aa.impl = new Impl(ti, ndim);
    else
        aa.resize(ndim);
```

Please verify my math.

Reply:
FYI, 20 million rounds up to 33 million as a power of 2. 20/33 < 4/5. So no rehashing should be happening.

Still a good idea to take into account the load factor.